src/heimdall/worker.cc

```diff
@@ -0,0 +1,903 @@
+  vtzero::index_value key_private_access_;
+  vtzero::index_value key_cash_only_toll_;
+  vtzero::index_value key_mode_change_;
+  vtzero::index_value key_named_intersection_;
+  vtzero::index_value key_is_transit_;
+  vtzero::index_value key_transition_count_;
+};
+
+} // anonymous namespace
+
+void heimdall_worker_t::ReadZoomConfig(const boost::property_tree::ptree& config) {
+  min_zoom_road_class_ = kDefaultMinZoomRoadClass;
+
+  auto tile_config = config.get_child_optional("tile");
+  if (tile_config) {
+    auto zoom_array = tile_config->get_child_optional("min_zoom_road_class");
+    if (zoom_array) {
+      size_t i = 0;
+      for (const auto& item : *zoom_array) {
+        if (i < kNumRoadClasses) {
+          min_zoom_road_class_[i] = item.second.get_value<uint32_t>();
+          ++i;
+        }
+      }
+      // Fill remaining values with defaults if array is too short
+      for (; i < kNumRoadClasses; ++i) {
+        min_zoom_road_class_[i] = kDefaultMinZoomRoadClass[i];
+      }
+    }
+  }
+
+  // Compute overall minimum zoom level (minimum of all road class zooms)
+  min_zoom_ = *std::min_element(min_zoom_road_class_.begin(), min_zoom_road_class_.end());
+}
+
+heimdall_worker_t::heimdall_worker_t(const boost::property_tree::ptree& config,
+                                     const std::shared_ptr<baldr::GraphReader>& graph_reader)
+    : config_(config), reader_(graph_reader),
+      candidate_query_(*graph_reader,
+                       TileHierarchy::levels().back().tiles.TileSize() / 10.0f,
+                       TileHierarchy::levels().back().tiles.TileSize() / 10.0f) {
+  ReadZoomConfig(config_);
+}
+
+std::unordered_set<GraphId>
+heimdall_worker_t::build_edges_layer(vtzero::tile_builder& tile,
+                                     const midgard::AABB2<midgard::PointLL>& bounds,
+                                     const std::unordered_set<baldr::GraphId>& edge_ids,
+                                     uint32_t z,
+                                     const TileProjection& projection,
+                                     bool return_shortcuts) {
+  using point_t = boost::geometry::model::d2::point_xy<double>;
+  using linestring_t = boost::geometry::model::linestring<point_t>;
+  using multi_linestring_t = boost::geometry::model::multi_linestring<linestring_t>;
+  using box_t = boost::geometry::model::box<point_t>;
+
+  // Create clip box with buffer to handle edges that cross boundaries
+  const box_t clip_box(point_t(-projection.tile_buffer, -projection.tile_buffer),
+                       point_t(projection.tile_extent + projection.tile_buffer,
+                               projection.tile_extent + projection.tile_buffer));
+
+  // Create edges layer builder
+  EdgesLayerBuilder edges_builder(tile);
+
+  // Collect unique nodes for the nodes layer
+  std::unordered_set<GraphId> unique_nodes;
+
+  for (const auto& edge_id : edge_ids) {
+    baldr::graph_tile_ptr edge_tile;
```

Comment:
could put it outside of this loop and maybe even sort edge_ids to minimize switching the tile